lib/ansible/modules/windows/win_domain_user.ps1

```diff
@@ -0,0 +1,278 @@
+$description = Get-AnsibleParam -obj $params -name "description" -type "str"
+$password = Get-AnsibleParam -obj $params -name "password" -type "str"
+$password_expired = Get-AnsibleParam -obj $params -name "password_expired" -type "bool"
+$password_never_expires = Get-AnsibleParam -obj $params -name "password_never_expires" -type "bool"
+$user_cannot_change_password = Get-AnsibleParam -obj $params -name "user_cannot_change_password" -type "bool"
+$account_locked = Get-AnsibleParam -obj $params -name "account_locked" -type "bool"
+$groups = Get-AnsibleParam -obj $params -name "groups"
+$enabled = Get-AnsibleParam -obj $params -name "enabled" -type "bool" -default $true
+$path = Get-AnsibleParam -obj $params -name "path" -type "str"
+$upn = Get-AnsibleParam -obj $params -name "upn" -type "str"
+
+# User informational parameters
+$user_info = @{
+    GivenName = Get-AnsibleParam -obj $params -name "firstname" -type "str"
+    Surname = Get-AnsibleParam -obj $params -name "surname" -type "str"
+    Company = Get-AnsibleParam -obj $params -name "company" -type "str"
+    EmailAddress = Get-AnsibleParam -obj $params -name "email" -type "str"
+    StreetAddress = Get-AnsibleParam -obj $params -name "street" -type "str"
+    City = Get-AnsibleParam -obj $params -name "city" -type "str"
+    State = Get-AnsibleParam -obj $params -name "state_province" -type "str"
+    PostalCode = Get-AnsibleParam -obj $params -name "postal_code" -type "str"
+    Country = Get-AnsibleParam -obj $params -name "country" -type "str"
+}
+
+# Parameter validation
+If ($account_locked -ne $null -and $account_locked) {
+    Fail-Json $result "account_locked must be set to 'no' if provided"
+}
+If (($password_expired -ne $null) -and ($password_never_expires -ne $null)) {
+    Fail-Json $result "password_expired and password_never_expires are mutually exclusive but have both been set"
+}
+
+try {
+    $user_obj = Get-ADUser -Identity $username -Properties *
+}
+catch {
+    $user_obj = $null
+}
+
+If ($state -eq 'present') {
+    # Ensure user exists
+    try {
+        $new_user = $false
+
+        # If the account does not exist, create it
+        If (-not $user_obj) {
+            If ($path -ne $null){
+                New-ADUser -Name $username -Path $path
+            }
+            Else {
+                New-ADUser -Name $username
+            }
+            $user_obj = Get-ADUser -Identity $username -Properties *
+            $new_user = $true
+            $result.changed = $true
+        }
+
+        # Set the password if required
+        If ($password -and (($new_user -and $update_password -eq "on_create") -or $update_password -eq "always")) {
+            $secure_password = ConvertTo-SecureString $password -AsPlainText -Force
+            Set-ADAccountPassword -Identity $username -Reset:$true -Confirm:$false -NewPassword $secure_password
+            $user_obj = Get-ADUser -Identity $username -Properties *
```

Comment:
That's what I did initially, but I found problems in testing. As accounts were updated, I ran into problems where later logic would fail. I ended up deciding it was better to refresh the state of the user object after every update.
 (Sorry, I know that's vague; I don't recall the specific issues I ran into off-hand.)